x509util/sctcheck/sctcheck.go

```diff
@@ -0,0 +1,242 @@
+		host += ":443"
+	}
+
+	glog.Infof("Retrieve certificate chain from TLS connection to %q", host)
+	dialer := net.Dialer{Timeout: hc.Timeout}
+	conn, err := cryptotls.DialWithDialer(&dialer, "tcp", host, &cryptotls.Config{InsecureSkipVerify: true})
+	if err != nil {
+		return nil, 0, 0, fmt.Errorf("failed to dial %q: %v", host, err)
+	}
+	defer conn.Close()
+
+	goChain := conn.ConnectionState().PeerCertificates
+	glog.Infof("Found chain of length %d", len(goChain))
+
+	// Convert base crypto/x509.Certificates to our forked x509.Certificate type.
+	chain := make([]*x509.Certificate, len(goChain))
+	for i, goCert := range goChain {
+		cert, err := x509.ParseCertificate(goCert.Raw)
+		if err != nil {
+			return nil, 0, 0, fmt.Errorf("failed to convert Go Certificate [%d]: %v", i, err)
+		}
+		chain[i] = cert
+	}
+
+	// Check externally-provided SCTs.
+	var valid, invalid int
+	scts := conn.ConnectionState().SignedCertificateTimestamps
+	if len(scts) > 0 {
+		merkleLeaf, err := ct.MerkleTreeLeafFromChain(chain, ct.X509LogEntryType, 0 /* timestamp added later */)
+		if err != nil {
+			glog.Errorf("Failed to build Merkle tree leaf: %v", err)
+			return chain, 0, len(scts), nil
+		}
+		for i, sctData := range scts {
+			subject := fmt.Sprintf("external SCT[%d]", i)
+			if checkSCT(ctx, lf, subject, merkleLeaf, sctData, ll, hc) {
+				valid++
+			} else {
+				invalid++
+			}
+
+		}
+	}
+
+	return chain, valid, invalid, nil
+}
+
+func checkSCT(ctx context.Context, liFactory logInfoFactory, subject string, merkleLeaf *ct.MerkleTreeLeaf, sctData []byte, ll *loglist.LogList, hc *http.Client) bool {
+	var sct ct.SignedCertificateTimestamp
+	if _, err := tls.Unmarshal(sctData, &sct); err != nil {
+		glog.Errorf("Failed to deserialize %s data: %v", subject, err)
+		return false
+	}
+	glog.Infof("Examine %s with timestamp: %d from logID: %x", subject, sct.Timestamp, sct.LogID.KeyID[:])
+	log := ll.FindLogByKeyHash(sct.LogID.KeyID)
+	if log == nil {
+		glog.Warningf("Unknown logID: %x, cannot validate %s", sct.LogID, subject)
+		return false
+	}
+	logInfo, err := liFactory(log, hc)
+	if err != nil {
+		glog.Errorf("Failed to build log info for %q log: %v", log.Description, err)
+		return false
+	}
+
+	glog.Infof("Validate %s against log %q...", subject, logInfo.Description)
+	if err := logInfo.VerifySCTSignature(sct, *merkleLeaf); err != nil {
+		glog.Errorf("Failed to verify %s signature from log %q: %v", subject, log.Description, err)
+	} else {
+		glog.Infof("Validate %s against log %q...validated", subject, log.Description)
+	}
+
+	glog.Infof("Check %s inclusion against log %q...", subject, log.Description)
+	index, err := logInfo.VerifyInclusion(ctx, *merkleLeaf, sct.Timestamp)
+	if err != nil {
+		age := time.Now().Sub(ct.TimestampToTime(sct.Timestamp))
+		if age < logInfo.MMD {
+			glog.Warningf("Failed to verify inclusion proof (%v) but %s timestamp is only %v old, less than log's MMD of %d seconds", err, subject, age, log.MaximumMergeDelay)
```

Comment:
Could you make the error messages printed by sctcheck and sctscan consistent if possible?

I only noticed this here as the error is in one message here whereas in two in sctscan, but this comment applies to all of the messages from the tools really - it's nice to have consistency :)

Reply:
I've tweaked a bit, but making them completely consistent might actually be confusing, because they're checking slightly different things:

 - `sctcheck` gets a live STH as it goes along, and the time delta effectively involves `time.Now()`
 - `sctscan` typically gets an STH at start-of-run and holds on to it, so the time delta may be against N hours ago (a full log scan can take many hours).

In general, the error messages are going to be different because these are very different tools:

 - `sctcheck` is pets: a few targets each hand-crafted on the command line
 - `sctscan` is cattle: millions of automatched targets. 

Detailed step-by-step logging makes sense for the former but not the latter.